core/src/main/java/io/questdb/griffin/SqlOptimiser.java

```diff
@@ -3801,6 +3801,126 @@ private ExpressionNode rewriteAggregate(ExpressionNode agg, QueryModel model) {
+             Core logic for issue #4231 will be applicable under two conditions
+             Condition1: QueryColumn contains FIRST/LAST/max/min function keyword
+             Condition2: QueryColumn for LAST/max keyword should be same as designatedTimestamp column
+             Core logic for changing query plan
+             - Add limit 1 via expressionNode to parent model
+             - Change model to select-choose from select-group-by
+             - Column alias by default is LAST/max if alias is not provided, replace with original column name
+             - Add order by clause to nested model
+             Call this recursively for nested models
+             */
+            if (optimisationType == 1 || optimisationType == 2) {
+                QueryModel newNestedModel = queryModelPool.next();
+                ExpressionNode lowerLimitNode = expressionNodePool.next();
+                lowerLimitNode.token = "1";
+                lowerLimitNode.type = 2;
+                parent.setLimit(lowerLimitNode, null);
+
+                //change model type to select-choose
+                parent.setSelectModelType(1);
+
+                //change ast params
+                ast.token = rhs;
+                ast.paramCount = 0;
+                ast.type = 4;
+
+                ExpressionNode newTimestampNode = expressionNodePool.next();
+                newTimestampNode.token = designatedTimestampColumn.toString();
+                if (optimisationType == 1)
+                    newNestedModel.addOrderBy(newTimestampNode, QueryModel.ORDER_DIRECTION_DESCENDING);
+                ObjList<QueryModel> joinModels = parent.getNestedModel().getJoinModels();
+                for (int i = 1, n = joinModels.size(); i < n; i++) {
+                    newNestedModel.addJoinModel(joinModels.getQuick(i));
+                }
+                newNestedModel.setTableNameExpr(nestedModel.getTableNameExpr());
+                newNestedModel.setModelType(nestedModel.getModelType());
+                newNestedModel.setTimestamp(nestedModel.getTimestamp());
+                newNestedModel.copyColumnsFrom(nestedModel, queryColumnPool, expressionNodePool);
+                parent.setNestedModel(newNestedModel);
+                modifiedNestedModel = newNestedModel;
+            }
+        }
+
+        final QueryModel union = parent.getUnionModel();
+        if (union != null) {
+            QueryModel modifiedUnionModel = rewriteGroupByForFirstLastMaxMinAggregateFunctions(union);
+            parent.setUnionModel(modifiedUnionModel);
+        }
+
+        ObjList<QueryModel> joinModels = parent.getJoinModels();
+        for (int i = 1, n = joinModels.size(); i < n; i++) {
+            QueryModel modifiedJoinModel = rewriteGroupByForFirstLastMaxMinAggregateFunctions(joinModels.getQuick(i));
+            joinModels.set(i, modifiedJoinModel);
+        }
+
+        rewriteGroupByForFirstLastMaxMinAggregateFunctions(modifiedNestedModel);
+        return parent;
```

Comment:
What's the point in returning the parent model if it's not replaced by a different `QueryModel`?

Reply:
@puzpuzpuz since we are not mutating the old model, so for instance if we are optimising any join model of parent model,  the modified model should be placed back into the join model array of parent model, but if we are not returning it from the call, we cannot have changed the model in the array object: 
Ref:- https://github.com/questdb/questdb/pull/4676/files#diff-cd500220c44e82328dfa75bdc225822a189a7b43b78ca0cd26ff962202edcf2bR3917

Reply:
The join model is passed as the `parent` to the `rewriteGroupByForFirstLastMaxMinAggregateFunctions(joinModels.getQuick(i))` call. So, since we don't replace parent model, but only its nested model, there is no point in returning the parent model from `rewriteGroupByForFirstLastMaxMinAggregateFunctions`. Am I missing something?